### ISSUE TITLE
Other similar approaches: static_assert(3)

### ISSUE THREAD

**Issue by alejandro-colomar** [jonnygrant/compile_assert#1]
https://github.com/jonnygrant/compile_assert/blob/9827107903f8e3b4160d5ae3056a1b47aa9de63b/README.md?plain=1#L8

I use `static_assert(3)` for similar purposes.  You need to be careful writing the expression, because static_assert() is always tested, even if it's within dead code.  For NULL checks I'm not sure it's viable, though.

See my definition of `ZUSTR2STP()`, a macro that wraps `zustr2stp()`, a function similar to `strncat(3)`, except that it does a simple copy instead of appending.

```c
$ grepc ZUSTR2STP .
./lib/zustr2stp.h:#define ZUSTR2STP(dst, src)                                                   \
({                                                                            \
	static_assert(!is_array(dst) || sizeof(dst) > SIZEOF_ARRAY(src), ""); \
                                                                              \
	zustr2stp(dst, src, NITEMS(src));                                     \
})
```

zustr2stp() is this:

```c
$ grepc -tfd zustr2stp .
./lib/zustr2stp.h:inline char *
zustr2stp(char *restrict dst, const char *restrict src, size_t sz)
{
	char  *p;

	p = mempcpy(dst, src, strnlen(src, sz));
	*p = '\0';

	return p;
}
```

and SIZEOF_ARRAY() is this:

```c
$ grepc SIZEOF_ARRAY .
./lib/sizeof.h:#define SIZEOF_ARRAY(a)  (sizeof(a) + must_be_array(a))
```

where must_be_array() is a static assertion that `a` is an array.

See the shadow-utils source code for more details:
<https://github.com/shadow-maint/shadow/blob/1c50a44db6bb02984fabbb482fea70ef2ca08fb2/lib/zustr2stp.h#L23>

**Comment by jonnygrant** [comment]
That's good. 

Does your approach have the ability to stop the compiler if it detects a NULL pointer?

I put the code from the static_assert man page in godbolt to see if is_array() could check a pointer, but doesn't seem to. Maybe I'm not using it correctly.

https://godbolt.org/z/jPedx7Yrj

**Comment by alejandro-colomar** [comment]
> That's good.
> 
> Does your approach have the ability to stop the compiler if it detects a NULL pointer?
> 
> I put the code from the static_assert man page in godbolt to see if is_array() could check a pointer, but doesn't seem to. Maybe I'm not using it correctly.

No.  What is_array() does is return `true` if the argument is an array, or `false` if it is not an array.

You'd need to write a `is_null_pointer_compiletime()` macro (something far from trivial), and then `static_assert(!is_null_pointer_compiletime(p));`.

You could check all GCC builtins, and maybe you find something that helps.

**Comment by jonnygrant** [comment]
> > That's good.
> > Does your approach have the ability to stop the compiler if it detects a NULL pointer?
> > I put the code from the static_assert man page in godbolt to see if is_array() could check a pointer, but doesn't seem to. Maybe I'm not using it correctly.
> 
> No. What is_array() does is return `true` if the argument is an array, or `false` if it is not an array.
> 
> You'd need to write a `is_null_pointer_compiletime()` macro (something far from trivial), and then `static_assert(!is_null_pointer_compiletime(p));`.

I tried to make one a few times, but didn't find a way in GCC to get it to work.
> 
> You could check all GCC builtins, and maybe you find something that helps.

Yes, those are good, like __builtin_constant_p() they do identify a few things, but that one needs a constant expression, but they can't analyse control flow, so can't detect bugs where a NULL is passed in.  So I came up with the compile_assert() macro, that does detect when NULL is passed in as an argument. It can be used to check out of bounds array indexing too etc

Maybe in another decade C and C++ will have more powerful static assertions. Certainly "Contracts" and "Profiles" will hopefully be standardised.

**Comment by alejandro-colomar** [comment]
Makes sense.  If you want to avoid some false positives, you could change compile_assert() to use a function `fail_linktime()` that will fail to link, instead of failing to compile.  That way, you give a chance to the LTO to optimize a few more cases.

I once used that trick, to write an `assert_unreachable()` macro that wouldn't be a hint to the compiler to optimize, but instead an assertion that the code is really unreachable.

**Comment by jonnygrant** [comment]
> Makes sense. If you want to avoid some false positives, you could change compile_assert() to use a function `fail_linktime()` that will fail to link, instead of failing to compile. That way, you give a chance to the LTO to optimize a few more cases.

That's a good idea. I did experiment with asm("compile_assert"); to make it fail to assemble the object file; before I settled on using __attribute__ ((error("foobar"))) to stop the compiler. But I think the missing symbol, didn't give me the useful file and line info that I wanted using __attribute__ ((error))

Could you give an example of the false-positive you are thinking might occur.

> 
> I once used that trick, to write an `assert_unreachable()` macro that wouldn't be a hint to the compiler to optimize, but instead an assertion that the code is really unreachable.

Sounds good 👍

**Comment by alejandro-colomar** [comment]
On Mon, Nov 27, 2023 at 02:12:04AM -0800, jonnygrant wrote:
> That's a good idea.  I did experiment with asm("compile_assert"); to
> make it fail to assemble the object file; before I settled on using
> __attribute__ ((error("foobar"))) to stop the compiler.  But I think
> the missing symbol, didn't give me the useful file and line info that
> I wanted using __attribute__ ((error))
> 
> Could you give an example of the false-positive you are thinking might occur.

A non-null that cannot be seen by the compiler due to being set in a
different translation unit than where it's being checked.  The linker
may be able to determine that it's non-null, since it sees all the TUs
at once, while the compiler may have thought that it was possible that
it were null, since it's opaque to it.

-- 
<https://www.alejandro-colomar.es/>

**Comment by alejandro-colomar** [comment]
> That's a good idea. I did experiment with asm("compile_assert"); to make it fail to assemble the object file; before I settled on using **attribute** ((error("foobar"))) to stop the compiler. But I think the missing symbol, didn't give me the useful file and line info that I wanted using **attribute** ((error))

Heh, it all depends on how you name your missing symbol.  ;)

```c
// Copyright 2023, Alejandro Colomar <alx@kernel.org>
// SPDX-License-Identifier: GPL-3.0-or-later OR LGPL-3.0-or-later


#define TOKENPASTE_(a, b)  a ## b
#define TOKENPASTE(a, b)   TOKENPASTE_(a, b)


#define assert_unreachable()  do                                              \
{                                                                             \
	extern void TOKENPASTE(assert_unreachable_in_line_, __LINE__)(void);  \
                                                                              \
	TOKENPASTE(assert_unreachable_in_line_, __LINE__)();                  \
} while (0)


int
main(void)
{
	assert_unreachable();
}
```

That won't give you the file name, an I don't think that's possible, but with the caller name and the line number, you should be covered in most cases (unless you define the caller function in several files, that is).

```sh
$ cc -Wall -Wextra -flto -fuse-linker-plugin ./test.c 
/usr/bin/ld: /tmp/ccHiJ0H1.ltrans0.ltrans.o: in function `main':
<artificial>:(.text+0x5): undefined reference to `assert_unreachable_in_line_20'
collect2: error: ld returned 1 exit status
```

----

I used the TOKENPASTE() macro from here: <https://stackoverflow.com/a/1597129/6872717>

**Comment by jonnygrant** [comment]
> > That's a good idea. I did experiment with asm("compile_assert"); to make it fail to assemble the object file; before I settled on using **attribute** ((error("foobar"))) to stop the compiler. But I think the missing symbol, didn't give me the useful file and line info that I wanted using **attribute** ((error))
> 
> Heh, it all depends on how you name your missing symbol. ;)
> 
> ```c
> // Copyright 2023, Alejandro Colomar <alx@kernel.org>
> // SPDX-License-Identifier: GPL-3.0-or-later OR LGPL-3.0-or-later
> 
> 
> #define TOKENPASTE_(a, b)  a ## b
> #define TOKENPASTE(a, b)   TOKENPASTE_(a, b)
> 
> 
> #define assert_unreachable()  do                                              \
> {                                                                             \
> 	extern void TOKENPASTE(assert_unreachable_in_line_, __LINE__)(void);  \
>                                                                               \
> 	TOKENPASTE(assert_unreachable_in_line_, __LINE__)();                  \
> } while (0)
> 
> 
> int
> main(void)
> {
> 	assert_unreachable();
> }
> ```
> 
> That won't give you the file name, an I don't think that's possible, but with the caller name and the line number, you should be covered in most cases (unless you define the caller function in several files, that is).
> 
> ```shell
> $ cc -Wall -Wextra -flto -fuse-linker-plugin ./test.c 
> /usr/bin/ld: /tmp/ccHiJ0H1.ltrans0.ltrans.o: in function `main':
> <artificial>:(.text+0x5): undefined reference to `assert_unreachable_in_line_20'
> collect2: error: ld returned 1 exit status
> ```
> 
> I used the TOKENPASTE() macro from here: https://stackoverflow.com/a/1597129/6872717

That looks like a useful approach.


I modified your example, to get the file:line number in a string, need to look that up per object file.

https://godbolt.org/z/jM8dxa37z

.LC0:
        .string "assert_unreachable_in_line_26:/app/example.cpp"
main:
        subq    $8, %rsp
        movl    $.LC0, %edi
        call    assert_unreachable_in_line_26(char const*)
        xorl    %eax, %eax
        addq    $8, %rsp
        ret

**Comment by alejandro-colomar** [comment]
> I modified your example, to get the file:line number in a string, need to look that up per object file.
> 
> https://godbolt.org/z/jM8dxa37z
> 
> .LC0: .string "assert_unreachable_in_line_26:/app/example.cpp" main: subq $8, %rsp movl $.LC0, %edi call assert_unreachable_in_line_26(char const*) xorl %eax, %eax addq $8, %rsp ret

Yeah, getting file:line is relatively easy.  The problem is that the linker won't show that in the error message.  It only mentions the undefined function, ignoring the string passed to it.

Although it's still useful if you inspect all the assembly files with `find -type f | grep '\.s$' | xargs grep assert_unreachable_in_line_26`, which will then show you the string you created.  That's an interesting debugging feature.

**Comment by alejandro-colomar** [comment]
So, to clean up a little bit:

```c
// Copyright 2023, Alejandro Colomar <alx@kernel.org>
// Copyright 2023, Jonny Grant <jg@jguk.org>
// SPDX-License-Identifier: GPL-3.0-or-later OR LGPL-3.0-or-later


#define STRINGIFY_(x)      # x
#define STRINGIFY(x)       STRINGIFY_(x)

#define TOKENPASTE_(a, b)  a ## b
#define TOKENPASTE(a, b)   TOKENPASTE_(a, b)


#define assert_unreachable()  do                                              \
{                                                                             \
	extern void TOKENPASTE(assert_unreachable_in_line_, __LINE__)(const char *); \
                                                                              \
	const char  *str;                                                     \
                                                                              \
	str = "assert_unreachable_in_line_" STRINGIFY(__LINE__) ":" __FILE__; \
	TOKENPASTE(assert_unreachable_in_line_, __LINE__)(str);               \
} while (0)


int
main(void)
{
	assert_unreachable();
}
```

While the linker doesn't report the file, we can get it easily:

```sh
$ cc -Wall -Wextra -flto -fuse-linker-plugin -S -O3 ./test.c
```
```sh
$ cc -Wall -Wextra -flto -fuse-linker-plugin -O3 ./test.s
/usr/bin/ld: /tmp/ccAG22Iy.ltrans0.ltrans.o: in function `main':
<artificial>:(.text.startup+0xc): undefined reference to `assert_unreachable_in_line_27'
collect2: error: ld returned 1 exit status
```
```sh
$ find * -type f  | grep '\.s$' | xargs grep -H assert_unreachable_in_line_27
test.s:	.string	"'assert_unreachable_in_line_27:./test.c"
test.s:	.string	"assert_unreachable_in_line_27"
```

GCC needs at least `-O2` to have the string in one line in the assembly file.
Clang provides the string in the assembly file under any optimization level:

```sh
$ find * -type f  | grep '\.s$' | xargs grep -H assert_unreachable_in_line_27:
test.s:@.str = private unnamed_addr constant [39 x i8] c"assert_unreachable_in_line_27:./test.c\00", align 1
```

**Comment by jonnygrant** [comment]
> On Mon, Nov 27, 2023 at 02:12:04AM -0800, jonnygrant wrote: That's a good idea. I did experiment with asm("compile_assert"); to make it fail to assemble the object file; before I settled on using __attribute__ ((error("foobar"))) to stop the compiler. But I think the missing symbol, didn't give me the useful file and line info that I wanted using __attribute__ ((error)) Could you give an example of the false-positive you are thinking might occur.
> A non-null that cannot be seen by the compiler due to being set in a different translation unit than where it's being checked. The linker may be able to determine that it's non-null, since it sees all the TUs at once, while the compiler may have thought that it was possible that it were null, since it's opaque to it.
> […](#)
> -- <https://www.alejandro-colomar.es/>

You're completely right, I don't know how much code optimization the linker does. There could be duplicate NULL checks on both sides of the API, not optimized away. For a safety critical system I accept that this does happen, as long as it doesn't show up in performance benchmarks as an issue I'm comfortable with it.

I take the view that safety critical systems must check their arguments they are passing to the_api(), but likewise the_api() in a safety critical system must also validate the values and pointers they receive. Human error is always a factor, so together there's more certainty that the system will be safe, the compiler may be used with optimization, so it might take out some of the checks as it sees fit too (but we also carefully check the assembler output too!). Sometimes we don't run with optimization turned on.

**Comment by jonnygrant** [comment]
I remember we spoke about these things a year ago. This is a way to get the line number 13 in the output message, rather than the linker

'compile_time_assertion_long_size_must_be_16_bytes_13'
https://godbolt.org/z/fca9P6afv

**Comment by jonnygrant** [comment]
It's a shame other two ways to output a message at compile time also don't have a way to concatenate and output FILE:LINE info. They all want pre-formatted string literals.

#pragma message
#warning


I've always avoided writing my own pre-processor that scans files for something like jonny_compile_assert() and updates a particular ("myfile.c:400") file and line number so the message is accurate when it recognises a particular warning string, although I know some projects do this, easy enough after a commit.

**Comment by alejandro-colomar** [comment]
Hi!

I know there's work in progress to add an `_Optional` qualifier to make sure NULLs don't sneak in where they shouldn't.

And for array bounds, I'm working on something myself: There's currently a patch from me for adding a `__countof__` operator to GCC which returns the number of elements of an array (just like the usual ARRAY_SIZE() macro).  It has already been accepted in ISO C2y, although we're in a fight for the name (it was accepted in C2y as `_Lengthof`, which I consider dangerous for being prone to confusion with the length of a string, and thus promoting dangerous off-by-one bugs).

When the patch for GCC is merged, I'll work in a second round of changes to GCC, which will make this operator support array parameters to functions:

```c
wchar_t *
my_wmemset(size_t n, wchar_t wcs[n], wchar_t wc)
{
        for (size_t i = 0; i < countof(wcs); i++)
                wcs[i] = wc;

        return wcs;
}
```

And there's another feature we're discussing at the moment, which is `[.identifier]` for specifying the number of elements before it's declared:

```c
wchar_t *
wmemset(wchar_t wcs[.n], wchar_t wc, size_t n)
{
        for (size_t i = 0; i < countof(wcs); i++)
                wcs[i] = wc;

        return wcs;
}
```

As you can see, the compiler would have full knowledge of the numbers of elements of arrays, and we suddenly have compile-time bounds checking in C with 0 overhead at run time, with minimal changes to the language, keeping it as simple as it always was.

**Comment by jonnygrant** [comment]
__countof__ would be really handy!

Would be good if the C standard could add support for function attributes 'error' and 'warning'
``
__attribute__((error("my message")))
``
``
__attribute__((warning("my message")))
``
in the same way noreturn  was added as _Noreturn in C11

so then we can do:
``
noreturn void terminate(int status)
{
    exit(status);
}
``

**Comment by alejandro-colomar** [comment]
> **countof** would be really handy!

The first patch for it is here:

<https://inbox.sourceware.org/gcc-patches/936f7945fae5ce7b5ca705a968af7c87474cc64c.1729622352.git.alx@kernel.org/#t>

This is the paper that was already merged to C2y earlier this month in the Minnesota meeting of WG14:

<https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3369.pdf>

And here's the current place where we're continuing the fight over the name, in case you want to participate:

<https://inbox.sourceware.org/gcc/lwvxugyioz7dmawjpq5pnqvwvne54fkhnwhzaxs6tviqgjvrmt@dhyc4mbkryqv/T/#m590055e314846e3c694715f03bfeb7e07de38195>

> Would be good if the C standard could add support for function attributes 'error' and 'warning' `__attribute__((error("my message")))` `__attribute__((warning("my message")))`

There's `[[deprecated]]` and `[[deprecated("reason")]]`.  You can use `-Wno-deprecated-declarations`, `-Wdeprecated-declarations -Wno-error=deprecated-declarations`, or `-Werror=deprecated-declarations` in GCC to control whether you want nothing, a warning, or an error.

**Comment by jonnygrant** [comment]
https://stackoverflow.com/questions/807244/c-compiler-asserts-how-to-implement

Alex

I recall you mentioned seeking the filename in the compile error:

someone got the filename into the token:

$ gcc -c demo.c
demo.c:32: error: size of array `assertion_failed_demo_c_32' is negative
$

Jonny

**Comment by alejandro-colomar** [comment]
Hi!

I was discussing today with someone in the C committee about being able to reject null pointers with code that can be written today.  I came up with something quite similar to your implementation (and the same name, except for an 'r'), even if I had completely forgotten about it.  :)

```c
alx@devuan:~/tmp$ cat memcpy.c 
#include <stddef.h>
#include <stdint.h>
#include <string.h>

#define compiler_assert(e)  do                                        \
{                                                                     \
	[[gnu::error("")]] extern void fail_(void);                   \
                                                                      \
	if (!(e))                                                     \
		fail_();                                              \
} while (0)

#define my_memcpy(dst, src, n)                                        \
({                                                                    \
	compiler_assert(dst != NULL);                                 \
	compiler_assert(src != NULL);                                 \
                                                                      \
	memcpy(dst, src, n);                                          \
})

void
f(char x[restrict 100], char y[restrict 100])
{
	my_memcpy(x, y, 10);

	if (x == NULL || y == NULL)
		return;

	my_memcpy(x, y, 100);
}
alx@devuan:~/tmp$ gcc -O1 memcpy.c 
memcpy.c: In function ‘f’:
memcpy.c:10:17: error: call to ‘fail_’ declared with attribute error: 
   10 |                 fail_();                                              \
      |                 ^~~~~~~
memcpy.c:15:9: note: in expansion of macro ‘compiler_assert’
   15 |         compiler_assert(dst != NULL);                                 \
      |         ^~~~~~~~~~~~~~~
memcpy.c:24:9: note: in expansion of macro ‘my_memcpy’
   24 |         my_memcpy(x, y, 10);
      |         ^~~~~~~~~
memcpy.c:10:17: error: call to ‘fail_’ declared with attribute error: 
   10 |                 fail_();                                              \
      |                 ^~~~~~~
memcpy.c:16:9: note: in expansion of macro ‘compiler_assert’
   16 |         compiler_assert(src != NULL);                                 \
      |         ^~~~~~~~~~~~~~~
memcpy.c:24:9: note: in expansion of macro ‘my_memcpy’
   24 |         my_memcpy(x, y, 10);
      |         ^~~~~~~~~
```

**Comment by jonnygrant** [comment]
Great. Did the C committee person wish to include something?

A compile_time_assert() is definitely a turbo charged static_assert(). I've used it to find overflow bugs and other issues in projects.

**Comment by alejandro-colomar** [comment]
> Great. Did the C committee person wish to include something?

Yes, he is proposing adapting C++26 contracts to C.
<https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3739.htm>

I'm opposed to it, as that can be done simpler with compiler_assert() and other simpler features.

> 
> A compile_time_assert() is definitely a turbo charged static_assert(). I've used it to find overflow bugs and other issues in projects.

Yup.

**Comment by jonnygrant** [comment]
> > Great. Did the C committee person wish to include something?
> 
> Yes, he is proposing adapting C++26 contracts to C. https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3739.htm
> 
> I'm opposed to it, as that can be done simpler with compiler_assert() and other simpler features.

Sounds right. My approach relies on an optimizer, as it needs dead code removal to work. C++26 contracts looks like it is just runtime, so that is more calls to abort()... which can't be used on safety critical systems that really **can't crash**.

> > A compile_time_assert() is definitely a turbo charged static_assert(). I've used it to find overflow bugs and other issues in projects.
> 
> Yup.
